### PR TITLE
Add password reset via email

### PR DESCRIPTION
**Pull request by vaasuu**

### CODE REVIEW

File: backend/controllers/password-reset.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    return res
+      .status(StatusCodes.BAD_REQUEST)
+      .json({ error: error.details[0].message });
+  }
+
+  const { email } = req.body;
+
+  const user = await users.findByEmail(email);
+  if (user === null) {
+    // user does not exist
```
Comment (JuhoBjn) [inline backend/controllers/password-reset.js:24]:
```suggestion

```
Comments should answer the question of why something is being done while the code tells you what is being done. Sometimes comments can also be used to explain complicated logic. I think this comment is not needed.

File: backend/controllers/password-reset.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    return res
+      .status(StatusCodes.NOT_FOUND)
+      .json({ error: "User with that email does not exist" });
+  }
+
+  const token = jwt.sign(
+    { id: user.id },
+    process.env.JWT_KEY,
+    { expiresIn: "30m" } // token expires in 30 minutes
```
Comment (JuhoBjn) [inline backend/controllers/password-reset.js:33]:
```suggestion
    { expiresIn: "30m" }
```
I think this comment is extra too.

File: backend/controllers/password-reset.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+  );
+
+  try {
+    await sendPasswordResetEmail(user, token); // send email
```
Comment (JuhoBjn) [inline backend/controllers/password-reset.js:37]:
```suggestion
    await sendPasswordResetEmail(user, token);
```
This comment also repeats the code.

File: backend/controllers/password-reset.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+      .status(StatusCodes.BAD_REQUEST)
+      .send({ error: error.details[0].message });
+  }
+
+  const { password, token } = req.body;
+
+  let decodedToken; // declare variable outside try-catch block
+
+  try {
+    // verify token
```
Comment (JuhoBjn) [inline backend/controllers/password-reset.js:76]:
```suggestion

```
This comment is extra.

File: backend/controllers/password-reset.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    if (!foundUser) {
+      return res
+        .status(StatusCodes.UNAUTHORIZED)
+        .send({ error: "Unauthorized" });
+    }
+
+    // bcrypt new password
+    const password_hash = await bcrypt.hash(password, 12);
+
+    // update password
```
Comment (JuhoBjn) [inline backend/controllers/password-reset.js:107]:
```suggestion

```
I think the code is quite self-explanatory here.

**Review by JuhoBjn — CHANGES_REQUESTED** [review]
Well-written code and all works. I've never sent emails from a server backend so I actually learned something new reading your code.
I left a couple suggestions regarding comments for you to check out.

**Comment by JuhoBjn** [comment]
I noticed Coveralls was failing one of the checks since test coverage had decreased 1,7%. I think that's a bit harsh for this project so I increased the coveage decrease threshold for failure to 10%.

**Comment by vaasuu** [comment]
> I noticed Coveralls was failing one of the checks since test coverage had decreased 1,7%. I think that's a bit harsh for this project so I increased the coveage decrease threshold for failure to 10%.

Noticed that too. Thanks for finding the knob to adjust it. After trying multiple hours to get mocks working correctly I just gave up trying to figure them out 🥲

File: backend/controllers/password-reset.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+  const { error } = schema.validate(req.body);
+  if (error) {
+    return res
+      .status(StatusCodes.BAD_REQUEST)
+      .send({ error: error.details[0].message });
+  }
+
+  const { password, token } = req.body;
+
+  let decodedToken; // declare variable outside try-catch block
```
Comment (vaasuu) [inline backend/controllers/password-reset.js:73]:
```suggestion
  let decodedToken;
```

File: backend/__tests__/password-reset.test.js [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+  });
+
+  describe("POST set-new-password", () => {
+    let user = {
+      id: "41284e8d-ee2b-4e15-b9af-296de2a9af8a" // Adam Administrator (test data)
+    };
+    let token = jwt.sign(
+      { id: user.id },
+      process.env.JWT_KEY,
+      { expiresIn: "30m" } // token expires in 30 minutes
```
Comment (vaasuu) [inline backend/__tests__/password-reset.test.js:96]:
```suggestion
      { expiresIn: "30m" }
```

**Review by JuhoBjn — APPROVED** [review]

**Pull request merged**